marshaling_rules.md

```diff
@@ -0,0 +1,148 @@
+# `R` <--> `Rust` marshaling rules
+
+This document outlines conversion and validation rules between `R` and `Rust` types.
+The aim is to maintain type safety on the `Rust` side without sacrificing usability on the `R` side.
+
+
+# Conversion and validation configuration
+[comment01](https://github.com/extendr/extendr/pull/261#discussion_r690303432)
+
+Configurations described below cover all possible cases. However, some of the combinations may not be needed in a real-life application. Suggested approach is to implement type-based configuration, which depends on the argument types defined in the user code. Here are some (top of the head) examples:
+
+- `Vec<i32>` triggers `NA` validation, altrep unfolding, and type coercion if compatible (so `1.0` converts to `1L`). Heavy on overhead and memory allocation, good for prototypes and testing things out.
+- `&[Rint]` is a responsible approach, which for array-based vectors of type `integer()` results in no allocations and no validation (zero overhead in the wrapper). **Unclear if type coercion should be performed**
+- `Integer` is an obscure wrapper of either `&[Rint]` or some `AltRepInt`. Acts as iterator with indexing capabilities, items are of type `Rint`, providing correct `NA` handling. Preferred way to handle vectors, mimics that of `{cpp11}`. **Unclear if type coercion should be performed**
+- (**Undecided**) `Numeric` is a discriminated union of `Integer | Real | Complex`. Accepts all numeric inputs, but leaves it up to the user to decipher what exactly was received from `R`. No runtime validation, no extra allocation, altreps remain altreps. 
```

Comment:
I think complex is not a numeric in R.

``` r
is.numeric(1+0i)
#> [1] FALSE
```

<sup>Created on 2021-08-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Fair enough, we can exclude it.

Reply:
However, even though it is not numeric, it seems to work nicely with basic `R` functions, including `mean`:

```R

mean(0:10 * 1i)
# > 0 + 5i
```

I would prefer to have `Numeric` include `Complex`.